circle_evolution/reporter.py

```diff
@@ -0,0 +1,85 @@
+"""Reporters for capturing events and notifying the user about them.
+
+You can make your own Reporter class or use the already implemented one's.
+Notice the abstract class before implementing your Reporter to see which
+functions should be implemented.
+"""
+from abc import ABC, abstractmethod
+
+import logging
+import logging.config
+
+
+class Reporter(ABC):
+    """Base Reporter class.
+
+    The Reporter is responsible for capturing particular events and sending
+    them for visualization. Please, use this class if you want to implement
+    your own Reporter.
+    """
+    def __init__(self):
+        """Initialization calls setup to configure Reporter"""
+        self.setup()
+
+    def setup(self):
+        """Function for configuring the reporter.
+
+        Some reporters may need configuring some internal parameters or even
+        creating objects to warmup. This function deals with this
+        """
+
+    @abstractmethod
+    def update(self, report):
+        """Receives report from subject.
+
+        This is the main function for reporting events. The Reporter receives a
+        report and have to deal with what to do with that. For Circle-Evolution
+        you can expect anything from strings to `numpy.array`.
+        """
```

Comment:
# First comment
I don't like the idea of having a single function to handle all commands.

My suggestion is to add many types of functions to deal with different events:

## Examples
For example, we can have an onStart method which is called when evolution is about to start.
We can have an onEnd method which is called when evolution ends.
We can have an onStop method, called when the user forcibly closes simulation.
onUpdate called when mutated species replaces the parent species.
onNextGen.... e.t.c

# Second comment
I also don't like the idea of having arguments passed as many types as this kind of code is prone to errors. I suggest using a class that holds information on the report.
For example, this class can hold the current generation number, current fitness, and best specie.

# Caveats

The problem with my first comment is that it can slow things down especially if using something like onNextGen... So tell me if there is a faster way to do it, or if you want to suggest something else.

The problem with my second comment is that if we pass in the species reference the programmer might modify it and it will modify the actual specie because it is only a reference. But again I believe whoever is programming should be wary, and know about this through the docs.

# Conclusion

Please tell me what you think about my two comments. If you have a better idea you may propose it to me.